lib/ask/ecto_types/schedule.ex

```diff
@@ -143,37 +143,91 @@ defmodule Ask.Schedule do
+  def last_window_ends_at(%{end_date: end_date, start_date: start_date} = schedule) do
+    backward = true
+    limit = start_date
+    from_date_time = Date.add(end_date, 1)
+    last_window_end = reversible_next_available_date_time(schedule, from_date_time, backward, limit)
+
+    if limit_exceed(last_window_end, backward, limit) do
+      raise ScheduleError, "last active window not found"
+    else
+      last_window_end
+    end
+  end
+
+  # Why do we need this reversible function? Because we need to calculate:
+  # 1. The start_time of the first active window (backward = false)
+  # 2. The end_time of the last active window (backward = true)
+  # And the logic for doing that is pretty much the same.
+  defp reversible_next_available_date_time(schedule, date_time, backward, limit) do
+    date_time = Timex.to_datetime(date_time, schedule.timezone)
+
     # TODO: Remove the necessity of converting this to erl dates
-    date_time = date_time
-    |> Timex.to_datetime(schedule.timezone)
+    {erl_date, erl_time} = Timex.to_erl(date_time)
+    {:ok, time} = Time.from_erl(erl_time)
 
-    {erl_date, erl_time} = date_time |> Timex.to_erl
+    selected_datetime = if day_of_week_available?(schedule, erl_date) && compare_time(schedule, time) == :inside do
+      date_time
+    else
+      selected_date = select_available_date(schedule, erl_date, time, backward, limit)
+      select_time_for_date(schedule, selected_date, backward)
+    end
+
+    selected_datetime
+    |> Timex.Timezone.convert("Etc/UTC")
+  end
 
-    {:ok, time} = erl_time
-    |> Time.from_erl
+  def select_time_for_date(schedule, erl_date, backward) do
+    if backward, do: at_end_time_erl(schedule, erl_date), else: at_start_time(schedule, erl_date)
+  end
 
-    # If this day is enabled in the schedule
-    date_time = if day_of_week_available?(schedule, erl_date) do
-      # Check if the time is inside the schedule time range
+  def select_available_date(schedule, erl_date, time, backward, limit) do
+    if day_of_week_available?(schedule, erl_date) do
       case compare_time(schedule, time) do
         :before ->
-          # If it's before the time range, move it to the beginning of the range
-          at_start_time(schedule, erl_date)
-        :inside ->
-          # If it's inside there's nothing to do
-          date_time
+          if backward do
+            next_available_date(schedule, erl_date, backward, limit)
+          else
+            erl_date
+          end
         :after ->
-          # If it's after the time range, find the next day
-          next_available_date_time_internal(schedule, erl_date)
+          if backward do
+            erl_date
+          else
+            next_available_date(schedule, erl_date, backward, limit)
+          end
       end
     else
-      # If the day is not enabled, find the next day
-      next_available_date_time_internal(schedule, erl_date)
+      next_available_date(schedule, erl_date, backward, limit)
```

Comment:
What do you think of something like this?

```elixir
def select_available_date(schedule, erl_date, time, backward, limit) do
    date_to_return = if day_of_week_available?(schedule, erl_date) do
      select_date_if_time(compare_time(schedule, time), backward)
    else
      :next_date
    end

    case date_to_return do
      :given_date -> erl_date
      :next_date -> next_available_date(schedule, erl_date, backward, limit)
    end
  end
```

where the new `select_date_if_time` fuction would be defined as follow

```elixir
  defp select_date_if_time(:before = _time_is, true = _backward), do: :next_date
  defp select_date_if_time(:before = _time_is, false = _backward), do: :given_date
  defp select_date_if_time(:after = _time_is, true = _backward), do: :given_date
  defp select_date_if_time(:after = _time_is, false = _backward), do: :next_date
```

I don't like the name for this new function, so If you find a better one, do change it please! 

PS: the tests run green with this 🙂